flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorFailureTest.java

```diff
@@ -160,8 +177,94 @@ public void testFailingCompletedCheckpointStoreAdd() throws Exception {
+
+        final CompletedCheckpointStore completedCheckpointStore =
+                new FailingCompletedCheckpointStore(
+                        (checkpoint, ignoredCleaner, ignoredPostCleanCallback) -> {
+                            throw failure;
+                        });
+
+        final AtomicInteger cleanupCallCount = new AtomicInteger(0);
+        final CheckpointCoordinator checkpointCoordinator =
+                new CheckpointCoordinatorBuilder()
+                        .setExecutionGraph(graph)
+                        .setCheckpointIDCounter(checkpointIDCounter)
+                        .setCheckpointsCleaner(
+                                new CheckpointsCleaner() {
+
+                                    private static final long serialVersionUID =
+                                            2029876992397573325L;
+
+                                    @Override
+                                    public void cleanCheckpointOnFailedStoring(
+                                            CompletedCheckpoint completedCheckpoint,
+                                            Executor executor) {
+                                        cleanupCallCount.incrementAndGet();
+                                        super.cleanCheckpointOnFailedStoring(
+                                                completedCheckpoint, executor);
+                                    }
+                                })
+                        .setCompletedCheckpointStore(completedCheckpointStore)
+                        .setTimer(manuallyTriggeredScheduledExecutor)
+                        .build();
+        checkpointCoordinator.triggerSavepoint(tmpFolder.newFolder().getAbsolutePath());
+        manuallyTriggeredScheduledExecutor.triggerAll();
+
+        try {
+            checkpointCoordinator.receiveAcknowledgeMessage(
+                    new AcknowledgeCheckpoint(
+                            graph.getJobID(), attemptId, checkpointIDCounter.getLast()),
+                    "unknown location");
+            fail("CheckpointException should have been thrown.");
+        } catch (CheckpointException e) {
+            assertThat(
+                    e.getCheckpointFailureReason(),
+                    is(CheckpointFailureReason.FINALIZE_CHECKPOINT_FAILURE));
+        }
+
+        assertThat(cleanupCallCount.get(), is(expectedCleanupCalls));
+    }
+
     private static final class FailingCompletedCheckpointStore implements CompletedCheckpointStore {
 
+        private final TriConsumerWithException<
+                        CompletedCheckpoint, CheckpointsCleaner, Runnable, Exception>
+                addCheckpointConsumer;
+
+        public FailingCompletedCheckpointStore(
+                TriConsumerWithException<
+                                CompletedCheckpoint, CheckpointsCleaner, Runnable, Exception>
+                        addCheckpointConsumer) {
+            this.addCheckpointConsumer = addCheckpointConsumer;
```

Comment:
I'm not sure whether the flexibility we get here with TriConsumer justifies the complexity added.
Wouldn't a field `private Exception failure;` passed to constructor sufficicient?

Reply:
You're right. It would be sufficient enough considering that `FailingCompletedCheckpointStore` is only used internally and does not need this extended flexibility. It was just a habit to generalize these kind of builders. But there is no need to do that considering that it's not used anywhere else. I'm gonna revert that change and go with your proposal.